### DISCUSSION TITLE
Manual/class-CompositeCollider2D

### DISCUSSION THREAD

**Discussion by giscus** [Developer-Notes-Extension/Unity-Notes#56] [Notes]
# Manual/class-CompositeCollider2D

https://docs.unity3d.com/Manual/class-CompositeCollider2D.html

**Comment by vertxxyz** [comment]
Here's a nice quote from melv, a physics programmer at Unity, from the Unity Discord:

> [Why does a Composite Collider require a rigidbody?]
Set the Rigidbody2D body-type to Static. This is identical to not adding a Rigidbody2D as the collider will get added to a static Rigidbody2D behind the scenes. It's not any overhead.
This is actually the core problem I had when I created the CompositeCollider2D. The problem was that colliders that don't have a Rigidbody2D are added to the static ground-body that is created behind the scenes. This means all such colliders are attached to this body that never moves. If you were to tell such a collider to use a composite then there's a few problems. The first is that we have no place for you to define the CompositeCollider2D for this behind-the-scenes body (it's not even a Rigidbody2D but a physics body). The other problem, assuming we could define a composite, was that all colliders like this would go to the same composite making them more and more expensive as you added more of them. When shapes are created in the physics system, they are created and attached to a body otherwise they don't work.
> 
> So the composite needs a body and it allows you to set that to static so you get identical behaviour and it allows you to control the scope of what gets blended with what. It also ensures that the colliders that are merged together are on the same rigidbody so they cannot move relative to each other which is how the physics wants to work as it moves bodies, not colliders. Moving colliders relative to the body is expensive. Moving bodies isn't because that's what the physics system is all about; moving bodies and updating transforms.
Adding a Rigidbody2D and setting to Static isn't an overhead. Go ahead and create as many as you like. They're, for the most part, ignored by the physics system. It's colliders, contacts and joints and solving them that costs.

You can follow [melv on twitter](https://twitter.com/melvmay).
